### ISSUE TITLE
Python , AI day1

### ISSUE THREAD

**Issue by Romangduck** [Romangduck/Day-1-Linux#10]
1)인공지능이란?

계산, 학습 등 인간의 지적능력을 컴퓨터를 통해 구현하는 기술

​

2)머신러닝과 딥러닝의 차이점

기본적으로 머신러닝과 딥러닝은 인공지능의 하위 분야로, 데이터 분석을 통해 패턴을 학습하고 결과를 예측하는 기술입니다. .

​

2).머신러닝

머신러닝은 일반적으로 상대적으로 적은 데이터를 사용하여 모델을 학습시키며, 각각의 입력 변수들에 대한 최적의 가중치(weight)를 찾습니다. 머신러닝은 대부분의 경우 수동으로 선택된 특성(feature)을 사용하여 입력을 처리하며, 일반적으로 선형 모델(linear model)이나 결정 트리(decision tree)와 같은 간단한 알고리즘을 사용합니다.

​

2)..딥러닝

반면에 딥러닝은 대규모 데이터를 사용하여 더 복잡한 모델을 학습시키며, 높은 수준의 추상화(abstraction)를 통해 입력 데이터의 특징(feature)을 자동으로 추출합니다. 딥러닝은 다층 신경망(multi-layer neural network)을 사용하여 입력 데이터를 처리하며, 비선형 함수를 사용하여 입력 데이터를 처리합니다.

따라서, 딥러닝은 머신러닝보다 더 복잡하고 정교한 모델을 구축할 수 있지만, 그만큼 대규모의 데이터와 연산 자원이 필요합니다. 또한, 딥러닝 모델을 학습시키는 데에는 시간과 비용이 더 많이 드는 경향이 있습니다.

딥러닝이란 ; 인간의 뉴런과 비슷한 인공신경망으로 정보 처리.

딥러닝은 사람이 추출한과정과 분류과정 모두를 합쳐서함.

머신러닝이란 ; 특정 부분을 스스로 학습해 성능 향상 

머신러닝은 데이터를 전처리하는 부분에서 사람이 특징을 추출

머신러닝이 딥러닝보다 더 큰 틀

인공지능쪽을 가려면 Tensorflow 자격증 따는것을 추천

​

ex).

y = wx + b → 최적의 값을 찾는것이 목표 ( y= 0 , 1 ) *최적의 매개변수 ( w , b ) 를 찾는것

w = weight ( 가중치 ) b = bios ( 편향 )

딥러닝에서는 가중치와 편향값을 찾으면 끝.

​

활성화함수는 인공신경망에서 입력값을 처리하여 출력값을 생성하는 함수입니다. 예를 들어, 어떤 사진을 입력으로 받아 고양이 사진인지 개 사진인지 판별하는 인공신경망이 있다고 가정해봅시다. 입력 이미지 데이터는 여러 개의 노드로 구성된 입력층으로 들어가게 되는데, 이때 각 노드는 이미지에서 한 개의 특징을 나타내며, 이 특징을 처리하여 다음 층으로 전달하게 됩니다. 이때 각 노드의 출력값을 결정하는 것이 활성화 함수입니다.

딥러닝은 다층 퍼센트론 ; 입력층 은닉층 출력층으로 구성 / /히든 레이어 구성이 중요한부분 

노드는 에지라고 하는 선으로 연결됨

​

Anaconda 설치 후 가상환경 생성

가상환경 생성 ( 프로젝트별로 다른 파이썬 환경 유지가능 )

conda create —name tf210 python=3.10.0 

2. 텐서플로우 설치 activate tf210 ( 가상환경 진입 ) conda install tensorflow ( 텐서플로우 설치 ) 

3. 가상환경 확인 conda env list ( 가상환경 리스트 확인 ) activate tf210 ( 가상환경 진입 ) python -V ( 파이썬 버전 확인 )

python import tensorflow as tf tf.version exit()

conda deactivate ( 가상환경 나오기 ) 

4. 라이브러리 설치 확인

conda list 

5. 라이브러리 설치 

conda install scikit-learn

conda install matplotlib

conda install numpy

conda install pandas 

6. 가상환경 지우기

 conda remove —name tf210 —all

-) 파이썬

​

Ctrl + / 하면 해당열 전부 주석처리 가능 !

​

딥러닝은 순서만 잘하자.

​

데이터 가져오기

모델구성하기 → 히든레이어 (은닉층 ) / ( input_dim=A ) *input_dim= A

컴파일,훈련

평가 , 예측

​

mse = 평균 제곱 오차 (loss)

예측값과 실제값의 차이를 제곱한 값들의 평균으로 계산됩니다. 즉, 예측값과 실제값의 차이를 제곱하여 평균화한 것입니다. MSE 값이 작을수록 모델의 예측 성능이 높다고 판단할 수 있습니다.

​

adam = 적응적 모멘트 추정 (optimizer)

기울기 갱신에 대한 효율성과 안정성을 높이는 데 주로 사용됩니다. Adam은 학습률의 초기값 설정에 민감하지 않고, 적응적으로 조절되기 때문에 학습의 불규칙성을 줄여줍니다. 또한, Adam은 모든 파라미터의 업데이트를 하나의 학습률로 수행하는 SGD보다 빠른 수렴을 보장할 수 있습니다.

​

model.complie(A) → A 컴파일

epochs = A → A번 훈련시키기 model.fit(x, y, epochs=1000) → 1000번 훈련

데이트양에 따라서 구성에 따라서 훈련량을 조절해줘야함

→ 모델이 학습 데이터에 지나치게 맞춰져서 새로운 데이터에 대한 일반화 성능이 저하되는 현상이다. 즉, 모델이 학습 데이터에 너무 딱 맞게 학습되어서 새로운 데이터에 대해서는 예측 능력이 떨어짐.

​

loss = model.evaluate(x,y) print(’loss : ‘ ,loss )

→ loss 가 0에 가까운 수를 찾아야함

​

result = model.predict([A])

print(’A의 예측값 : ‘ , result)

​

arrary() → 행와 열을 맞춰주는 함수

![image](https://user-images.githubusercontent.com/129341526/236809895-78d96168-fe23-49a4-9578-1d6fecc0cf2e.png)
![image](https://user-images.githubusercontent.com/129341526/236809920-f92e3147-8e94-44ab-b782-91098185339f.png)

​

히든레이어 → 뉴런값이 크다고해서 좋은것이 아님. 적절한게 좋다.

​

batch → 더 많은 출력 값

배치 사이즈(batch size)란, 한 번의 업데이트에 사용되는 샘플의 개수 배치 사이즈가 32인 경우, 32개의 샘플을 모아서 한 번의 업데이트를 수행함 ▪ 작은 배치 사이즈는 모델이 더 자주 업데이트되므로 노이즈가 많은 경향이 있으며, 더 많은 학습 반복이 필요 ▪ 큰 배치 사이즈는 안정적인 경향이 있으나 메모리 문제와 학습 시간이 길어지는 문제가 발생할 수 있음 ▪ 딥러닝에서 배치 사이즈는 학습시에 지정되며 일반적으로 32, 64, 128 등의 2의 거듭제곱 수를 사용함 ▪ 배치 사이즈는 모델의 크기와 데이터셋의 크기에 따라 적절하게 조정되어야 함

데이터가 어떻게 생겼는지 파악하는것 ; 그것이 우선적으로 판단되야함

​

인공지능 개념 정리 - 머신러닝 , 딥러닝

​

인공지능이란?

계산, 학습 등 인간의 지적능력을 컴퓨터를 통해 구현하는 기술

​

머신러닝

머신러닝은 일반적으로 상대적으로 적은 데이터를 사용하여 모델을 학습시키며, 각각의 입력 변수들에 대한 최적의 가중치(weight)를 찾습니다. 머신러닝은 대부분의 경우 수동으로 선택된 특성(feature)을 사용하여 입력을 처리하며, 일반적으로 선형 모델(linear model)이나 결정 트리(decision tree)와 같은 간단한 알고리즘을 사용합니다.

​

머신러닝의 특징

특정 부분을 스스로 학습해 성능 향상 

머신러닝은 데이터를 전처리하는 부분에서 사람이 특징을 추출

머신러닝이 딥러닝보다 더 큰 틀

​

딥러닝

반면에 딥러닝은 대규모 데이터를 사용하여 더 복잡한 모델을 학습시키며, 높은 수준의 추상화(abstraction)를 통해 입력 데이터의 특징(feature)을 자동으로 추출합니다. 딥러닝은 다층 신경망(multi-layer neural network)을 사용하여 입력 데이터를 처리하며, 비선형 함수를 사용하여 입력 데이터를 처리합니다.

​

딥러닝의 특징

인간의 뉴런과 비슷한 인공신경망으로 정보 처리.

딥러닝은 사람이 추출한과정과 분류과정 모두를 합쳐서함.

​

​

머신러닝과 딥러닝의 차이점

​

딥러닝은 머신러닝보다 더 복잡하고 정교한 모델을 구축할 수 있지만, 그만큼 대규모의 데이터와 연산 자원이 필요합니다. 또한, 딥러닝 모델을 학습시키는 데에는 시간과 비용이 더 많이 드는 경향이 있습니다.

​

​

​

2. 퍼셉트론 (Perceptron) 

​

딥러닝의 일종으로 인공 신경망의 한 종류이다.

뉴런의 모델을 모방하여 입력층,출력층으로 구성한 모델이다.

구성요소 : 입력값, 가중치 , 활성화함수 , 출력값(예측값)

​

-> 뉴런으로 들어오는 각 입력값에 곱해지는 가중치 값을 자동적으로 학습하는 알고리즘

머신러닝의 지도학습이나 분류의 맥락에서 볼 때 , 하나의 샘플이 어떤 클래스에 속해 있는지 예측하는데 사용될 수 있다.

다수의 트레이닝 데이터를 이용하여 일조으이 지도 학습을 수행하는 알고리즘이라 볼 수 있다.

입력층과 출력층 사이를 중간측이라 부르는데 이는 노드 또는 뉴런이라 부르기도한다.

중간층이 하나의 노드로 구성되어 중간층과 출력층의 구분이 없는 구조를 단층 퍼셉트론이라 부른다.

​

3. 다층 퍼셉트론 (Multi-Layer Perceptron: MLP) 

단층 퍼셉트론으로 AND 연산에 대해서는 학습이 가능하지만, XOR에 대해서는 불가능하다는 한계점이 있다. 이를 극복하기 위해 입력층과 출력층 사이에 두개 이상의 중간층을 두어 비선형적으로 분리되는 데이터에 대해서도 학습이 가능하도록 다층 퍼셉트론이 고안됬다.

입력층과 출력층 사이에는 은닉층이 있고, 여러개의 은닉층이 있는 인공 신경망을 심층 신경망이라 부르며, 심층 신경망을 학습하기 위해 고안된 특별한 알고리즘들을 딥러닝이라 부른다.

다층 퍼셉트론은 은닉층과 출력층에 존재하는 활성 함수가 단층 퍼셉트론과 달리 여러개이므로 , 이에 따른 가중치도 여러개다. 

다층 퍼셉트론의 한계점 : 단층 퍼셉트론은 출력층의 결과값을 우리가 확보한 실제값과 비교하여 오차가 최소가 되도록 가중치를 업데이트하고 결정하면 되지만 , 다층 퍼셉트론은 은닉층의 출력값에 대한 맞다 틀리다하는 기준이 없다.

이를 해결 하기위해서, 출력층에서 발생하는 오차값을 이용하여 은닉층으로 역전파시켜 은닉층에서 발생하는 오차값에  따라 은닉층의 가중치를 업데이트함.

​

4. 옵티마이저 (Optimizer)

​

최적화 (Optimization)는 손실 함수의 결과값을 최소화하는 모델의 가중치를 찾는것을 의미한다.

이것의 알고리즘이 Optimizer 이다.

옵티마이저는 학습 데이터셋을 이용하여 모델을 학습할때 데이터의 실제 결과와 모델이 예측한 결과를 기반으로 잘 줄일 수 있게 만들어주는 역할을 한다. 

현재 가장 많이 사용하는 옵티마이저는 Adam이다.
![image](https://user-images.githubusercontent.com/129341526/236809753-0a8a1928-6d3f-4e68-8a37-53022fb19bb5.png)

5. 학습률 (learning rate)

(복습요망)

학습률이란 매 가중치에 대해 구해진 기울기 값을 얼마나 경사 하강법에 적용할지를 결정하는 하이퍼 파라미터이다.

학습률이 작다면 손실이 최적인 가중치를 찾는데 오랜 시간이 걸릴 것이며, 학습률이 너무 크다면 최적점을 무질서하게 이탈할 수도 있습니다.

​

울기를 계산한 후에 기울기 값을 기존 가중치(기울기)에 더하거나 빼서 가중치를 업데이트하는데 이때 계산된 기울기에 미리 설정해둔 학습률을 곱해서 나온 값으로 가중치를 업데이트하도록 한다. 예를들어 내가 계산한 기울기 값이 +6이어서 기존의 가중치 값이 5이면 가중치의 업데이트 이후에는 -1이 된다. 하지만 학습률을 0.1로 설정하게 되면 +6 * 0.1로 0.6이 되고 이 때의 업데이트 결과는 5 - 0.6으로 4.4가 된다.

​

학습률이 너무 작으면 (0.00001), 5 - 0.00006 = 4.99994 가 되므로 가중치의 변화가 거의 없어 학습으로 가중치가 제대로 갱신도 되지 않고 끝나버리게 된다. 따라서 학습률을 어떻게 잡는지는 중요한 이슈이다. 

6. 경사하강법 (Gradient Descent) 

(복습요망)

경사하강법은 1차 근삿값 발견용 최적화 알고리즘이다, 함수의 기울기를 이용해 x 의 값을 어디로 옮길 때 최솟값을 갖는지 알아보는 방법이다. 

기울기가 양수인경우 x 가 작아지는 방향으로 옮겨야하고, 기울기가 음수인경우 x 가 커지는 방향으로 옮겨야한다.

7. 손실함수 (Loss Function) 

​

실제값과 예측값의 차이(오차)를 비교하는 지표 

값이 낮을수록 학습이 잘 된 것이며, 정답과 알고리즘 출력을 비교할 때 사용 

최적의 매개변수(가중치와 편향)을 탐색할 때 손실 함수를 가능한 작게 하는 값을 찾음 

MSE(Mean Squared Error) : 평균 제곱 오류 출력결과와 데이터 차이 제곱의 평균으로 정답과 오답의 모든 확률 고려 

CEE(Cross Entropy Error) : 교차 엔트로피 오차 실체 정답의 확률만을 고려한 손실 함수(훈련데이터의 정답 혹은 확률이 0 또는 1

​

8. 활성화 함수 (Activation Function) – Sigmoid, ReLU, Softmax 

​

활성화 함수(Activation Function)란 인공 신경망(Neural Network)에서 사용되는 은닉층(Hidden Layer)을 활성화하기 위한 함수입니다.

가중치(Weight)와 편향(Bias)을 갱신(Update)하기 위해 사용되는 주요한 함수입니다.

​

예를 들어 나의 출근 시간을 예측하기 위한 모델(Model)을 구현한다고 가정해보겠습니다.

입력값이 일어난 시간(x1), 기상 상태(x2)라면, 기상 상태(x2)값 보다는 일어난 시간(x1)이 더 영향력이 크다는 것을 직관적으로 알 수 있습니다.

그러므로 연산 과정에서 일어난 시간(x1)은 더 많이 활성화(Activate)되어야 하며, 기상 상태(x2)는 비교적 비활성화(Deactivate) 되어야 합니다.

활성화 함수(Activation Function)는 비선형(Non-linear) 구조를 가져 역전파 과정에서 미분값을 통해 학습이 진행될 수 있게 합니다.

만약, 활성화 함수가 선형(Linear) 구조라면, 미분 과정에서 항상 상수가 나오게 되므로 학습이 진행되지 않습니다.

​

활성화함수의 종류

(복습요망)

이진 활성화 함수 (Binary step activation function)

선형 활성화 함수 (Linear activation function)

비선형 활성화 함수 (Non-linear activation function)

**Comment by Romangduck** [comment]
https://blog.naver.com/rrlagusejr/223096554270
